More torch fixes

honno opened the pull request:
* Remove use of `_UndefinedStub` in `dtype_helpers.py` 
  * This is mighty useful as currently we have two classes of the dtype helpers—those that consume stubs and those who don't, which is hard to keep track. This unifies the approach of the helpers so they are only constructed with available dtypes.
* Resolves #179 
* Resolves #178 by introducing a `hh.reject_overflow()` context manager that deals with non-`OverflowError` overflow errors (i.e. torch's `RuntimeError: ...overflow...`)
* Resolves #180
* Resolves #186
* Resolves #166

asmeurer commented:
There's also this error. I can't tell if it's caused by this or something else:

```
_______________________________________________________________________________ test_prod _______________________________________________________________________________

    @given(
>       x=xps.arrays(
            dtype=xps.numeric_dtypes(),
            shape=hh.shapes(min_side=1),
            elements={"allow_nan": False},
        ),
        data=st.data(),
    )

array_api_tests/test_statistical_functions.py:112:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
array_api_tests/test_statistical_functions.py:143: in test_prod
    d_m, d_M = dh.dtype_ranges[default_dtype]
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = EqualityMapping({torch.int8: MinMax(min=-128, max=127), torch.int16: MinMax(min=-32768, max=32767), torch.int32: MinMa...86e+38, max=3.4028234663852886e+38), torch.float64: MinMax(min=-1.7976931348623157e+308, max=1.7976931348623157e+308)})
key = None

    def __getitem__(self, key):
        for k, v in self._key_value_pairs:
            if key == k:
                return v
        else:
>           raise KeyError(f"{key!r} not found")
E           KeyError: 'None not found'
E           Falsifying example: test_prod(
E               x=tensor(0, dtype=torch.uint8),
E               data=data(...),
E           )
E           Draw 1 (kw): {}

array_api_tests/dtype_helpers.py:83: KeyError
```

asmeurer commented:
I think that's it. Every other test failure seems to be either an issue with torch or an unrelated test suite problem which I've reported elsewhere.

asmeurer commented:
Not that it's that important, but can we address https://github.com/data-apis/array-api-tests/pull/181#issuecomment-1528045938 here?

honno commented:
> `full` uses the default complex dtype by default, which is complex64 for torch, so it shouldn't generate fill values outside of that dtype unless dtype=complex128 explicitly.

> There's also this error. I can't tell if it's caused by this or something else:

Fixed and fixed—thanks for catching these. I folded in #182 as it addresses the former issue. <sup>Whilst its nice to keep PRs atomic I think we're not keeping track of multiple PRs at once, so mono PRs with at least a general cohesion is probably better.</sup>

> Not that it's that important, but can we address [#181 (comment)](https://github.com/data-apis/array-api-tests/pull/181#issuecomment-1528045938) here?

Already addressed!

honno commented:
(I'll merge this soon if no review)
